src/main/java/org/apache/mesos/offer/ResourceUtils.java

```
@@ -211,6 +218,49 @@ public static ExecutorInfo clearResourceIds(ExecutorInfo executorInfo) {
             .build();
   }
 
+  public static boolean areDifferent(
+          ResourceSpecification oldResourceSpecification,
+          ResourceSpecification newResourceSpecification) {
+
+    Protos.Value oldValue = oldResourceSpecification.getValue();
+    Protos.Value newValue = newResourceSpecification.getValue();
+    if (!ValueUtils.equal(oldValue, newValue)) {
+      LOGGER.info("Values are different.");
+      return true;
+    }
+
+    if (areDifferent(oldResourceSpecification.getRole(), newResourceSpecification.getRole())) {
+      LOGGER.info("Roles are different.");
+      return true;
+    }
+
+    if (areDifferent(oldResourceSpecification.getPrincipal(), newResourceSpecification.getPrincipal())) {
+      LOGGER.info("Principals are different.");
+      return true;
+    }
+
+    return false;
+  }
+
+  private static boolean areDifferent(String string1, String string2) {
+    return !(string1 == null ? string2 == null : string1.equals(string2));
+  }
+
+  public static Protos.Resource updateResource(Protos.Resource resource, ResourceSpecification resourceSpecification) {
+    Protos.Resource.Builder builder = Protos.Resource.newBuilder(resource);
+    switch (resource.getType()) {
+      case SCALAR:
+        return builder.setScalar(resourceSpecification.getValue().getScalar()).build();
+      case RANGES:
+        return builder.setRanges(resourceSpecification.getValue().getRanges()).build();
+      case SET:
+        return builder.setSet(resourceSpecification.getValue().getSet()).build();
+      default:
+        LOGGER.error("Encountered unexpected Value type: " + resource.getType());
```

Throw IllegalArgumentException here? this feels like a good time to fail early